includes/Story_Renderer.php

```diff
@@ -325,6 +335,75 @@ protected function replace_body_end_tag( $content ) {
 		return str_replace( '</body>', $output . '</body>', $content );
 	}
 
+	/**
+	 * Get the configuration object to use.
+	 *
+	 * @return Configuration Optimizer configuration to use.
+	 */
+	private function get_optimizer_configuration() {
+		$transformers = [
+			AmpSchemaOrgMetadata::class,
+			AmpRuntimeCss::class,
+			TransformedIdentifier::class,
+			ReorderHead::class,
+		];
+
+		$configuration = [ Configuration::KEY_TRANSFORMERS => $transformers ];
+		$config        = new Configuration( $configuration );
+		$config->registerConfigurationClass(
+			AmpSchemaOrgMetadata::class,
+			AmpSchemaOrgMetadataConfiguration::class
+		);
+
+		return $config;
+	}
+
+	/**
+	 * Optimizes the resulting markup using AMP Optimizer if available.
+	 *
+	 * @param string $content Story markup.
+	 *
+	 * @return string Filtered content.
+	 */
+	protected function optimize_markup( $content ) {
+		if ( ! class_exists( '\AmpProject\Optimizer\TransformationEngine' ) ) {
+			return $content;
+		}
+
+		$enable_optimizer = true;
+
+		/** This filter is documented in amp-wp/includes/class-amp-theme-support.php */
+		$enable_optimizer = apply_filters( 'amp_enable_optimizer', $enable_optimizer );
+
+		if ( ! $enable_optimizer ) {
+			return $content;
+		}
+
+		// Needed for \AmpProject\Optimizer\Transformer\AmpRuntimeCss.
+		$content = str_replace( '</head>', '<style amp-runtime></style></head>', $content );
```

Comment:
The `TransformedIdentifier` transformer is what is is responsible for flagging a document as being transformed which would then allow `style[amp-runtime]`, which you are including. So that seems OK. But injecting the element in this way seems not ideal.

Flagging a document as being `transformed` is normally bound up with a document also going through SSR. Hence the `AmpRuntimeCss`, `ServerSideRendering`, and `TransformedIdentifier` form a pair: https://github.com/ampproject/amp-wp/blob/8d5f29f1f22e6623789dca250901cff0a559e590/includes/class-amp-theme-support.php#L2182-L2194

The `ServerSideRendering` transformer adds the `style[amp-runtime]` element. However, why is this not the responsibility of the `AmpRuntimeCss` transformer? 

@schlessera Should [this logic](https://github.com/ampproject/amp-wp/blob/8d5f29f1f22e6623789dca250901cff0a559e590/lib/optimizer/src/Transformer/ServerSideRendering.php#L201-L204) be moved from `ServerSideRendering` to `AmpRuntimeCss`?

```php
// Emit the amp-runtime marker to indicate that we're applying server side rendering in the document.
$ampRuntimeMarker = $document->createElement(Tag::STYLE);
$ampRuntimeMarker->setAttribute(Attribute::AMP_RUNTIME, '');
$document->head->insertBefore($ampRuntimeMarker, $document->head->hasChildNodes() ? $document->head->firstChild : null);
```

Reply:
The `ServerSideRendering` does all the checks to see if SSR can be done, and then tags the runtime to denote that it should be inlined.

The `AmpRuntimeCss` is responsible for inlining (or linking) the runtime CSS, but only does so if the SSR gave its okay.

That models the way it is also handled in the `amp-toolbox` implementation.

If we move the generation of the element into the `AmpRuntimeCss` transformer, we'd need another element/attribute of some sort to provide the information whether inlining should happen or not.

Reply:
Given that stories can't do SSR, should the AmpRuntimeCss be done?

Reply:
> Given that stories can't do SSR, should the AmpRuntimeCss be done?

I don't know too much about AMP SSR, so if you can give me some guidance that would be great :-)